packages/tools/webpack-fluid-loader/src/protocol.ts

```diff
@@ -0,0 +1,213 @@
+        }
+    }
+
+    private broadcastClients() {
+        for (const [member, client] of this.members) {
+            this.emit("addMember", member, client);
+        }
+    }
+
+    private connectClient(clientId: string, client: ISequencedClient) {
+        this.members.set(clientId, client);
+        this.broadcastClients();
+    }
+
+    connectLocalClient(clientId: string, sequenceNumber: number) {
+        const sequencedClient: ISequencedClient = {
+            sequenceNumber,
+            client: {
+                mode: "write",
+                permission: [],
+                user: {
+                    id: clientId,
+                },
+                scopes: [
+                    "doc:read",
+                    "doc:write",
+                ],
+                details: {
+                    capabilities: {
+                        interactive: true,
+                    },
+                },
+                timestamp: Date.now(),
+            },
+        };
+
+        this.connectClient(clientId, sequencedClient);
+    }
+
+    disconnectLocalClient(clientId: string) {
+        if (this.members.get(clientId) !== undefined) {
+            this.emit("removeMember", clientId);
+            this.members.delete(clientId);
+        }
+    }
+
+    getMembers(): Map<string, ISequencedClient> {
+        return new Map<string, ISequencedClient>(this.members);
+    }
+
+    getMember(clientId: string): ISequencedClient | undefined {
+        return this.members.get(clientId);
+    }
+
+    dispose(_error?: Error | undefined): void {
+        this.disposed = true;
+    }
+
+    async propose(key: string, value: any): Promise<void> {
+        this.proposals.set(key, value);
+        return new Promise<void>(() => { });
+    }
+
+    has(key: string): boolean {
+        return this.proposals.get(key) !== undefined;
+    }
+
+    get(key: string): any {
+        return this.proposals.get(key);
+    }
+
+    add(snapshot: IQuorumSnapshot): LocalQuorum {
+        for (const pair of snapshot.values) {
+            const proposal = pair[1];
+            this.proposals.set(proposal.key, proposal.value);
+        }
+
+        return this;
+    }
+}
+
+class EmptyProtocolHandler implements IProtocolHandler {
+    constructor(
+        public readonly audience: IAudienceOwner,
+        public readonly quorum: LocalQuorum,
+        public readonly attributes: IDocumentAttributes,
+        public readonly initialSnapshot: IQuorumSnapshot,
+    ) { }
+
+    public snapshot(): IQuorumSnapshot {
+        return this.initialSnapshot;
+    }
+
+    processMessage(message: ISequencedDocumentMessage, _local: boolean): IProcessMessageResult {
+        const emptyResult = {};
+        const clientId = message.clientId;
+        if (this.quorum.getMember(clientId) !== undefined) {
+            return emptyResult;
+        }
+
+        this.quorum.connectLocalClient(clientId, message.sequenceNumber);
```

Comment:
I think the better name for connectLocalClient would be ensureClientExists(), or something like that

Reply:
sure, but this is not a class we're going to be actually shipping. This is kind of an example/model.